### PR TITLE
Add parsers for sequence files

### PR DESCRIPTION
**Pull request by fgrunewald**
This PR introduces some handy parsers for linear sequence files. This gives a bit more flexibility to specify exotic patterns of sequences without having to go through either the .json syntax or using the -seq flag. This solves issue #191 and #23. In addition it simplifies generation of cyclic ssDNA and future dsDNA. 

Parsers available: 

-  ~~csv~~
- space delimited txt
- fasta
- ig

@jan-stevens Are there any other useful DNA formats we should support?

### CODE REVIEW

File: polyply/src/simple_seq_parsers.py [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+    be any special character.
+    """
+    monomers = []
+    for line in lines:
+        for resname in line.strip().split(delimiter):
+            monomers.append(resname)
+    seq_graph =  _monomers_to_linear_nx_graph(monomers)
+    return seq_graph
+
+def parse_plain(lines, DNA=False, RNA=False):
```
Comment (fgrunewald) [inline polyply/src/simple_seq_parsers.py:74]: @pckroon: I'm implementing some parsers for simple sequence files. Would you pass the lines of the file to the parser or the handle and then open the file in the parser itself?

**Comment by fgrunewald** [comment]
@jan-stevens thanks for the suggestions. Let's make the oxDNA parser it's own thing since it generates more complex topologies as you write. This PR first deals with simple sequences and together with the DNA completion should work fine for most stuff

File: polyply/src/simple_seq_parsers.py (hunk 1 above)
Comment (jbarnoud) [reply polyply/src/simple_seq_parsers.py:65]: I usually pass an iterator of lines (opened file or list of lines, the function does not need to know). If I need to pass a file name for convenience, I have another function that is just a very thin wrapper around the first function.

Comment (fgrunewald) [reply polyply/src/simple_seq_parsers.py:65]: pretty much like it is now?

Comment (jbarnoud) [reply polyply/src/simple_seq_parsers.py:65]: yes

**Review by jbarnoud — COMMENTED** [review]
Looking good overall. I just have "a few" comments, nothing but suggestions, though.

I notice that, when reading fasta or ig files, you only rely on the comments to figure out if you are dealing with DNA or RNA. It means you need to edit your input files for polyply specifically in most cases. It could be beneficial to have the parsers take the molecule type as optional argument so you can pass it as CLI argument.

File: polyply/src/meta_molecule.py [hunk 2]
```diff
@@ -239,31 +256,16 @@ def from_monomer_seq_linear(cls, force_field, monomers, mol_name):
         return meta_mol_graph
 
     @classmethod
-    def from_json(cls, force_field, json_file, mol_name):
+    def from_sequence_file(cls, force_field, file_handle, mol_name):
         """
-        Constructs a :class::`MetaMolecule` from a json file
-        using the networkx json package.
+        Generate a meta_molecule from known sequence file parsers. Currently
+        plain, fasta, IG, as well as csv and txt are known for linear sequences
```
[the remark this answers is not in the document]
Comment (fgrunewald) [reply polyply/src/meta_molecule.py:262]: So you'd just remove the supported formats, and basically it will throw an error if it doesn't know them?

File: polyply/src/meta_molecule.py [hunk 3]
```diff
@@ -239,31 +256,16 @@ def from_monomer_seq_linear(cls, force_field, monomers, mol_name):
...
-                except TypeError:
-                    msg = "Couldn't add 1 to node. Either provide resids or use integers as node keys."
-                    raise IOError(msg)
-            graph.add_node(node, **attrs)
+        extension = file_handle.suffix.casefold()[1:]
+        if extension in cls.parsers:
+            graph = cls.parsers[extension](file_handle)
 
-        graph.add_edges_from(init_graph.edges(data=True))
         meta_mol = cls(graph, force_field=force_field, mol_name=mol_name)
```
[the remark this answers is not in the document]
Comment (fgrunewald) [reply polyply/src/meta_molecule.py:269]: Good catch: I'll add an error message

File: polyply/src/simple_seq_parsers.py [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+        lines = file_.readlines()
+
+    monomers = []
+    for line in lines:
+        for resname in line.strip().split(delimiter):
+            monomers.append(resname)
+    seq_graph =  _monomers_to_linear_nx_graph(monomers)
+    return seq_graph
+
+parse_csv = partial(parse_plain_delimited, delimiter=',')
```
[the remark this answers is not in the document]
Comment (fgrunewald) [reply polyply/src/simple_seq_parsers.py:62]: can you strip the quotation marks to allow for both and/or mixed format?

File: polyply/src/meta_molecule.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (jbarnoud) [reply polyply/src/meta_molecule.py:262]: The issue is that this docstring is too far from https://github.com/marrink-lab/polyply_1.0/pull/192/files/9f4498aefb55481bbfc2d362de7171dd7c26ccec#diff-86937c389af9e11f56c0f8e490b59355e347f000f1269a8d3ea3c9244abbdffdR92

I would put the documentation somewhere users actually look in a way that is visible enough that devs will know it is there when they add a format; or I would move the dict linked above in the function, so a dev adding a format has the docstring on screen and think about updating it.

File: polyply/src/simple_seq_parsers.py [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+    for idx, line in enumerate(lines):
+        clean_line, comment = split_comments(line)
+        comments.append(comment)
+        if clean_line:
+            if clean_line[-1] == '1' or clean_line[-1] == '2':
+                ter_char = clean_line[-1]
+                clean_line = clean_line[:-1]
+                clean_lines.append(clean_line)
+                DNA, RNA = _identify_nucleotypes(comments)
+                seq_graph = parse_plain(clean_lines[1:], DNA=DNA, RNA=RNA)
```
[the remark this answers is not in the document]
Comment (fgrunewald) [reply polyply/src/simple_seq_parsers.py:178]: Because according to the format specification the first non-comment line will be the name and then comes the actual sequence. Probably should add a comment here

File: polyply/src/simple_seq_parsers.py [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+    for line in lines:
+        for resname in line.strip().split(delimiter):
+            monomers.append(resname)
+    seq_graph =  _monomers_to_linear_nx_graph(monomers)
+    return seq_graph
+
+parse_csv = partial(parse_plain_delimited, delimiter=',')
+parse_txt = parse_plain_delimited
+
+def parse_plain(lines, DNA=False, RNA=False):
```
[the remark this answers is not in the document]
Comment (fgrunewald) [reply polyply/src/simple_seq_parsers.py:65]: I don't understand how enums would fix this, but I've never used them before. My logic was that running the _identify_nucleotypes before takes care to filter out invalid input.

File: polyply/tests/test_meta_molecule.py [hunk 7]
```diff
@@ -121,24 +122,24 @@ def test_from_monomer_seq_linear(monomers, edges, nodes, attrs):
...
         assert set(meta_mol.edges) == set(edges)
 
-    @staticmethod
-    def test_from_json_error_resid():
-        ff = vermouth.forcefield.ForceField(name='test_ff')
-        name = "test"
-        file_name = TEST_DATA + "/json/fail.json"
-        with pytest.raises(IOError):
-            meta_mol = MetaMolecule.from_json(ff, file_name, name)
+ #  @staticmethod
```
[the remark this answers is not in the document]
Comment (fgrunewald) [reply polyply/tests/test_meta_molecule.py:136]: Good catch. That's actually a leftover. I think this error is raised from a different function now but I wanted to make sure to still have that tested. Seems it slipped my mind. I'll check if the error is covered then remove it.

File: polyply/src/simple_seq_parsers.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (jbarnoud) [reply polyply/src/simple_seq_parsers.py:62]:
```python
if len(resname) >=2 and resname[0] == resname[-1] and resname[0] in ("'", '"'):
    resname = resname[1:-1]
```

This would work assuming the delimiter and spaces are invalid characters in resnames. Otherwise, you will need to account for escape characters as well and you are better off with importing a third party csv parser.

File: polyply/src/simple_seq_parsers.py [hunk 8]
```diff
@@ -0,0 +1,236 @@
...
+    Read in json file.
+    """
+    with open(filehandle) as file_:
+        data = json.load(file_)
+
+    init_json_graph = nx.Graph(json_graph.node_link_graph(data))
+    # the nodes in the inital graph are not ordered, when no resid
+    # is given this can create issues. So we reorder the node based
+    # on the node key, which HAS to be numerical.
+    seq_graph = nx.Graph(node_dict_factory=OrderedDict)
```
[the remark this answers is not in the document]
Comment (fgrunewald) [reply polyply/src/simple_seq_parsers.py:231]: So far I've had the opinion of better safe than sorry, as python has changed it's mind several times on this issue. I didn't know they finally committed to it. But I'd still leave it. Changing version support is something I'd like to do with the next major version but I'll open an issue about this.

File: polyply/src/simple_seq_parsers.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (fgrunewald) [reply polyply/src/simple_seq_parsers.py:62]: and here I thought csv was simple; I'll see if there is a good lightweight csv parser

File: polyply/src/simple_seq_parsers.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (jbarnoud) [reply polyply/src/simple_seq_parsers.py:65]: I realised that afterwards. If that is the idea, then `parse_plain` shouldn't be exposed to the users. Maybe add a comment and rename the function `_parse_plain`.

File: polyply/src/simple_seq_parsers.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (jbarnoud) [reply polyply/src/simple_seq_parsers.py:178]: I did not realise it was a list of lines and not a list of characters. My bad.

File: polyply/src/meta_molecule.py [hunk 9]
```diff
@@ -101,6 +105,19 @@ def __init__(self, *args, **kwargs):
...
+        # present. All algroithm relay on proper resids
+        for node in self.nodes:
+            if "resid" not in self.nodes[node]:
+                LOGGER.warning("Node {} has no resid. Setting resid to {} + 1.", node, node)
+
+                try:
+                    self.nodes[node]["resid"] = node + 1
+                except TypeError:
+                    msg = "Couldn't add 1 to node. Either provide resids or use integers as node keys."
+                    raise IOError(msg)
```
[the remark this answers is not in the document]
Comment (fgrunewald) [reply polyply/src/meta_molecule.py:118]: test has been added

File: polyply/src/meta_molecule.py [hunk 10]
```diff
@@ -239,31 +256,16 @@ def from_monomer_seq_linear(cls, force_field, monomers, mol_name):
         return meta_mol_graph
 
     @classmethod
-    def from_json(cls, force_field, json_file, mol_name):
+    def from_sequence_file(cls, force_field, file_handle, mol_name):
```
[the remark this answers is not in the document]
Comment (fgrunewald) [reply polyply/src/meta_molecule.py]: name has been changed

File: polyply/src/meta_molecule.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (fgrunewald) [reply polyply/src/meta_molecule.py]: I changed the docstring and put a pointer to the parser dict. Also an unkwon file format will raise an error now.

File: polyply/src/simple_seq_parsers.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (fgrunewald) [reply polyply/src/simple_seq_parsers.py:65]: So I had a loot at the python standard library for csv and they simple implement different language types like unix, excel etc. I'm not sure I want to do this, because that seems pretty complicated as you'd have to set the language type through CLI and cannot identify it automatically. I'm thinking to remove csv support. @jbarnoud do you have a better idea?

File: polyply/src/simple_seq_parsers.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (fgrunewald) [reply polyply/src/simple_seq_parsers.py]: True sometimes I forget about the private function synatx convention. I added a comment and renamed the function accordingly.

File: polyply/src/simple_seq_parsers.py [hunk 11]
```diff
@@ -0,0 +1,236 @@
...
+            elif token in ONE_LETTER_RNA and RNA:
+                resname = ONE_LETTER_RNA[token]
+            else:
+                msg = f"Cannot find one letter residue match for {token}"
+                raise IOError(msg)
+
+            monomers.append(resname)
+
+    # make sure to set the defaults for the DNA terminals
+    if DNA:
```
[the remark this answers is not in the document]
Comment (fgrunewald) [reply polyply/src/simple_seq_parsers.py]: added the RNA support. This was a leftover since previously it would also parse AA one letter sequences. Since I dropped support for that there is no need to check if RNA or DNA is set. One can just always adding the ending by default.

File: polyply/src/simple_seq_parsers.py [hunk 12]
```diff
@@ -0,0 +1,236 @@
...
+    ----------
+    filehandle: :class:`pathlib.Path`
+    """
+    with open(filehandle) as file_:
+        lines = file_.readlines()
+
+    clean_lines = []
+    comments = []
+    idx = 0
+    for idx, line in enumerate(lines):
```
[the remark this answers is not in the document]
Comment (fgrunewald) [reply polyply/src/simple_seq_parsers.py:176]: Done! It reads better now indeed and also the error is a good addition.

File: polyply/src/simple_seq_parsers.py [hunk 13]
```diff
@@ -0,0 +1,236 @@
...
+                    seq_graph.add_edge(0, nnodes-1)
+                    seq_graph.edges[(0, nnodes-1)]["linktype"] = "circle"
+                    seq_graph.nodes[0]["resname"] = seq_graph.nodes[0]["resname"][:-1]
+                    seq_graph.nodes[nnodes-1]["resname"] = seq_graph.nodes[nnodes-1]["resname"][:-1]
+                break
+            else:
+                clean_lines.append(clean_line)
+
+    if idx < len(lines):
+        LOGGER.warning("Found more than 1 sequence. Will only use the first one.")
```
[the remark this answers is not in the document]
Comment (fgrunewald) [reply polyply/src/simple_seq_parsers.py:202]: I adjusted the warning by adding another condition that the next line may not be a comment.

File: polyply/src/simple_seq_parsers.py [hunk 14]
```diff
@@ -0,0 +1,236 @@
...
+
+    return seq_graph
```
[the remark this answers is not in the document]
Comment (fgrunewald) [reply polyply/src/simple_seq_parsers.py:204]: Now an error should be raised if that happens

File: polyply/src/simple_seq_parsers.py [hunk 15]
```diff
@@ -0,0 +1,236 @@
...
+    resname attribute.
+    """
+    seq_graph = nx.Graph()
+    mon_range = range(0, len(monomers))
+    seq_graph.add_edges_from(zip(mon_range[:-1], mon_range[1:]))
+    nx.set_node_attributes(seq_graph, {node: resname for node, resname in zip(seq_graph.nodes, monomers)}, "resname")
+    nx.set_node_attributes(seq_graph, {node: node+1 for node in seq_graph.nodes}, "resid")
+    return seq_graph
+
+def parse_plain_delimited(filehandle, delimiter=" "):
```
[the remark this answers is not in the document]
Comment (fgrunewald) [reply polyply/src/simple_seq_parsers.py:50]: Yes I understand but I'm always torn between the two things. Personally I use it a lot to quickly create meta-molecules from file names/paths, which is very convenient like in an ipython terminal. It's fine to open the file and then pass the lines but that's some extra lines I've to add and I'm lazy. I'll ponder this over the weekend see if I can make up my mind. Vermouth is also split on this. Like the pdb reader takes a file path itp/ff readers take lines.

File: polyply/src/meta_molecule.py [hunk 16]
```diff
@@ -90,6 +89,11 @@ class MetaMolecule(nx.Graph):
     """
 
     node_dict_factory = OrderedDict
+    parsers = {"csv": parse_csv,
```
Comment (jbarnoud) [inline polyply/src/meta_molecule.py:92]: Don't forget to remove.

File: polyply/src/simple_seq_parsers.py [hunk 17]
```diff
@@ -0,0 +1,245 @@
...
+            if clean_line[-1] == '1' or clean_line[-1] == '2':
+                ter_char = clean_line[-1]
+                clean_line = clean_line[:-1]
+                clean_lines.append(clean_line)
+                break
+            else:
+                clean_lines.append(clean_line)
+    else:
+        msg = "The sequence is not complete, it does not end with 1 or 2."
+        raise FileFormatError(msg)
```
Comment (jbarnoud) [inline polyply/src/simple_seq_parsers.py:188]: You miss a test with a sequence in IG format that does not end with 1 or 2 to make sure you handle the error correctly.

File: polyply/src/simple_seq_parsers.py [hunk 18]
```diff
@@ -0,0 +1,245 @@
...
+    seq_graph = _parse_plain(clean_lines[1:], DNA=DNA, RNA=RNA)
+
+    if ter_char == '2':
+        nnodes = len(seq_graph.nodes)
+        seq_graph.add_edge(0, nnodes-1)
+        seq_graph.edges[(0, nnodes-1)]["linktype"] = "circle"
+        seq_graph.nodes[0]["resname"] = seq_graph.nodes[0]["resname"][:-1]
+        seq_graph.nodes[nnodes-1]["resname"] = seq_graph.nodes[nnodes-1]["resname"][:-1]
+
+    if idx < len(lines) - 1 and not lines[idx+1].startswith(";"):
```
Comment (jbarnoud) [inline polyply/src/simple_seq_parsers.py:200]: If you have more than 1 sequence, the next line after the first sequence may very much be a comment to tell what that other is. Instead of pilling up the checks, I would rephrase the error message to be more speculative. "There may be more than one sequence in the file. We will only use the first one."

File: polyply/src/simple_seq_parsers.py [hunk 19]
```diff
@@ -0,0 +1,245 @@
...
+        LOGGER.warning("Found more than 1 sequence. Will only use the first one.")
```
Comment (jbarnoud) [inline polyply/src/simple_seq_parsers.py:201]: Also, you need to test a file with content after the sequence.

**Review by jbarnoud — APPROVED** [review]
Looks ready to me.
